### PR TITLE
Add asdict and update to evented dataclass

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR add:
 
1) general event `values_updated` for update or any object properties
2) `as_dict` and `update` methods
3) deprecation events support (for future refactors)

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/utils/events/dataclass.py
Comment (tlambert03) [inline napari/utils/events/dataclass.py]: this will definitely save time when the arrays are different sizes... unfortunately, we now can't use `or` in the conditional, or we might get an `AttributeError` when accessing `.size` (if one isn't an array). However, we still have to know if _either_ of them is an array :/

Also, let's use `np.array_equal` rather than `np.any` to avoid the deprecation warnings that @sofroniewn saw.  

we also need to know if they are dask arrays, and not just numpy arrays (but bail out short of comparison, and just assume they are different)... otherwise `bool(before != after)` will still fail.

I did some tests and it definitely does help to check if they're arrays first (as opposed to simply using `np.array_equal` directly).  Here's the fastest, most general solution I could come up with (if not the prettiest):
```python
def is_different(a, b):
    if hasattr(a, 'size'):
        if hasattr(b, 'size'):
            return not np.array_equal(a, b) if a.size == b.size else True
        return True
    if hasattr(b, 'size'):
        return True
    return bool(a != b)
```

that will raise a FutureWarning if either is a dask array... so feel free to keep modifying and playing around with the idea.

Comment (Czaki) [reply napari/utils/events/dataclass.py]: I  add this because I got FutureWarnning for compare array of different size when running test set. 
Now I see that we should add some test to pass tuple, list and dask array. And fix it.

Comment (Czaki) [reply napari/utils/events/dataclass.py]: @tlambert03 @sofroniewn I try to think a little about this. And I'm not sure if the signal should not be emitted if there are two different representations of the same array.

If value is `[1, 2, 3]` and senewtted value is `numpy.array([1, 2, 3])` It may be new semanic value (some object could treat list and numpy array different?

Comment (tlambert03) [reply napari/utils/events/dataclass.py]: yeah, I was thinking about that too.  Since this is a very low-level object that will be used in many different ways, I think we should err on the side of emission (and fast evaluation), rather than try to get clever about avoiding signal emission for extremely similar (but technically different) objects.  There are [other ways](https://github.com/napari/napari/blob/2508efc85a1141cf6ac9d422f4dc9ce6a69d6c1f/docs/source/explanations/dataclass.md#gettersetter-side-effects--conditional-event-logic) for users to prevent event emission if it's really important.

Comment (Czaki) [reply napari/utils/events/dataclass.py]: The current comparison also will fail when having dict with nested numpy arrays. 
I found such package: https://pypi.org/project/deepdiff/

What did you think about adding this dependency? Something like:
```
def compare(v1, v2):
    if isinstance(v1, da.core.Array) or isinstance(v2, da.core.Array):
        return False
    diff = DeepDiff(v1, v2, ignore_numeric_type_changes=True, max_diffs=1)
    return diff == {}
```

Comment (tlambert03) [reply napari/utils/events/dataclass.py]: as mentioned, I'm hesitant to add nearly _any_ overhead in this particular function (since it can be added in `_on_<name>_set` if careful diffing is required).  But, if you want to do some tests with `%%timeit` and can demonstrate that it doesn't get any slower while also meeting more edge cases, i guess we could consider the dependency

Comment (tlambert03) [reply napari/utils/events/dataclass.py]: (I'm also fine just having a try/catch and saying "yes they're different" if the comparison fails)

File: napari/utils/events/dataclass.py
Comment (Czaki) [inline napari/utils/events/dataclass.py:184]: I found that this comparison could return false positive. for example when before is `1` and after is `np.array([1,1])`

File: napari/utils/events/dataclass.py
Comment (Czaki) [inline napari/utils/events/dataclass.py]: Not sure if this implementation is correct.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: Ready

File: napari/utils/events/dataclass.py
Comment (tlambert03) [reply napari/utils/events/dataclass.py:184]: That's one reason why I suggested using no.array_equal in a previous comment, but also, because this use of np.any and piecewise comparison will throw a warning

Comment (tlambert03) [reply napari/utils/events/dataclass.py:184]: https://github.com/napari/napari/pull/1966#discussion_r532115845

Comment (Czaki) [reply napari/utils/events/dataclass.py:184]: Ok

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: This function should probably live in its own utils file, maybe `events/utils.py`.

I don't love that this now depends on dask and numpy, but maybe it's unavoidable for now ...

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: What is the reason for the `values_updated` event?

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: I don't love adding this to each `evented_dataclass`, is there a way to add deprecated events after the object is made?

I've also found the current syntax https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L692 that we added in #1582 limiting as sometimes there is not a drop in replacement provided as is, and so would probably prefer something that just let me pass the entire custom warning. Not sure if that's starting to sound like the following then https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L603

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: It looks like the individual events will get emitted when they are set. Is this just an extra event to let you know the whole dataclass is updated?

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: we can probably drop `_get_state` if we have `as_dict`

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: Curious what people think of these names `as_dict` and `update_from_dict`. Alternatives could be just `from_dict` or `asdict` and `fromdict` (to match the dataclass syntax)?

I like `as_dict`. `update_from_dict` might be a bit long, but at least it is clear, so I'm ok with it too. I imagine we'll use these quite a lot though so good to talk it over @jni

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py:224]: Do we know of cases where this will happen?

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: I'm just looking at some of the complexity introduced into this class to handle the case where event emission should be blocked when the input value is the same as the output value. We used this pattern a lot in the codebase before the evented_dataclass, which is why I think it was added, but maybe we should step back and consider dropping it, i.e. if you ever try and set then we will emit event, even if value hasn't changed.

There might be some places in the codebase now where we will want to add if statements like this outside the dataclass, or maybe we can clean up some code to prevent us setting things that shouldn't be set.

Curious what @tlambert03 @jni @Czaki think of this idea?

**Review by sofroniewn — COMMENTED** [review]
I've given this some review @Czaki. Some of us are teaching at the ImageJ and friends conference right now so might be slow to respond.

File: napari/utils/events/dataclass.py
Comment (tlambert03) [reply napari/utils/events/dataclass.py]: I'm so so on the addition of that event as well.  I'd tend towards waiting to find a use case where just connecting directly to specific events is insufficient before we start adding events that might be useful someday.

File: napari/utils/events/dataclass.py
Comment (tlambert03) [reply napari/utils/events/dataclass.py]: I do kind of wish that `dataclasses` had used `as_dict` instead of `asdict` in the first place, but now that they _have_ chosen `asdict`, I'm slightly inclined to stick with the standard library phrasing.

`fromdict` has sort of changed here, since as @Czaki pointed out, "from_dict" can be done just by `Class(**dict)`, and what's been implemented here really is more of an `update`.  It'd be nice to use `update` alone (following `dict.update`?), but since this will be on _every_ `evented_dataclass`, I don't think we want to use up that API.

(I'd be fine without it at all too)

File: napari/utils/events/dataclass.py
Comment (tlambert03) [reply napari/utils/events/dataclass.py]: > if you ever try and set then we will emit event, even if value hasn't changed.

I'm definitely open to that.  i agree it's getting hairier than I wanted for something going on every attribute of every class

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: > It'd be nice to use update alone (following dict.update?), but since this will be on every evented_dataclass, I don't think we want to use up that API.

I don't know, I think it might be ok - it's an update, like `dict.update`, I kinda like it. I'd be supportive of `asdict` and `update` as the two names

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: > I'm definitely open to that. i agree it's getting hairier than I wanted for something going on every attribute of every class

Ok i might look into what the consequences of dropping this entirely would be in a new PR. Would be good to understand before making a decision. If we drop then that keeps these extra dependencies out of the class, which could really one day live with our event system in it's own standalone `events` library (if we so wanted)

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py:224]: compare of two dictionaries with NumPy arrays as values, comparison of two list/tuples with NumPy arrays as elements. 
It is covered by tests.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: > This function should probably live in its own utils file, maybe `events/utils.py`.

I'm not an enthusiast of this. Its do only shallow comparison. Maybe It should be changed to a private method. 

> I don't love that this now depends on dask and numpy, but maybe it's unavoidable for now ...

Its already (earlier) depends on NumPy. I only add dask to avoid expensive calculations.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: For example camera synchronization. 
Maybe refactor VispyCamera which could be a little simplified. 

For me, it covers all cases when we would like to use `as_dict` Then we do not need to know (on the level of implementation) exactly the list of parameters (signals). 

Similar need we could observe in many places where there is near the same code to change different parameters of the object.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: > I don't love adding this to each `evented_dataclass`, is there a way to add deprecated events after the object is made?

To achieve this `EmitterGroup` needs to be updated. 

> I've also found the current syntax [https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L692](https://github.com/napari/napari/blob/master/napari/utils/events/event.py?rgh-link-date=2020-12-01T03%3A43%3A13Z#L692) that we added in #1582 limiting as sometimes there is not a drop in replacement provided as is, and so would probably prefer something that just let me pass the entire custom warning. Not sure if that's starting to sound like the following then [https://github.com/napari/napari/blob/master/napari/utils/events/event.py#L603](https://github.com/napari/napari/blob/master/napari/utils/events/event.py?rgh-link-date=2020-12-01T03%3A43%3A13Z#L603)


My idea for #1582 is to add a very simple mechanism. I think that WarningEmitter could be used in a more complicated state. But it does not support fast connection to already exists code and needs many more lines of code to be used. 

A possible refactor is to allow pass tuple (namedtuple) of arguments instead of the name of the signal, which tuple will contain information like signal name, custom error message, etc. But it will produce a more complicated code. 

As evented dataclass starts to be widely used inside napari and napari still sometimes fast changes it should start to provide a simple mechanism of backward compattybility. 

`values_updated` is part of my proposition of such API. simple deprecation method (which allows to not write many lines of code to inform the user about deprecation) also. Another option may be some __post_init__ method. but inheritance is problematic here.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: I especially use the longer name to avoid name, collition. But `update` should be fine if it could have dict, or instance of the class as an argument?  

For me `update` look like a method that will accept class instance.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: As I understand I'm against this. In many places in codebases signals are used to synchronize state (I think that Camer ecosystem could be an example). So if a guardian is not provided then It could come in an infinity loop. 

Also experience from another libs is that set of the same value will not call callbacks/emit signals.

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: > As evented dataclass starts to be widely used inside napari and napari still sometimes fast changes it should start to provide a simple mechanism of backward compattybility.
>
> values_updated is part of my proposition of such API. simple deprecation method (which allows to not write many lines of code to inform the user about deprecation) also. Another option may be some post_init method. but inheritance is problematic here.

I think we will already need to add custom getters to do a deprecation which is very easy with the `Property` syntax, as we'll need to emit warnings when people just access the property even if they don't try and connect events to it. Given that we'll have to do custom things like that when we deprecated I'd really prefer not to add this special property "deprecated_events" to all our dataclasses right now.

I also think this is something we can solve when we deprecate our first property and not in a PR to add an `update` method to the dataclass, which is unrelated.

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: The normal events keep on getting emitted to though, so you have really have to decide now do you listen to individual events or to the `update` event as doing both will cause to double responses.

Also should we then check the dict for equality to see if any properties change and event actually needs to be emitted (if we still do that check for individual properties?).

You can also connect to `dataclass.events` if you want an event each time any property is changed, which is different, but now we have three different ways to listen to events which I think is too much.

Like @tlambert03 said, we could add this at some point if we found that we wanted to start using it internally (and camera might be the right place to see), but I'm hesitant to expand the API right now unless we absolutely must

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: > In many places in codebases signals are used to synchronize state (I think that Camer ecosystem could be an example). So if a guardian is not provided then It could come in an infinity loop.

We'd definitely have to fix these, but that might improve the overall codebase!

> Also experience from another libs is that set of the same value will not call callbacks/emit signals.

Ok that's good to know. Is that the case with Qt?

Curious what @jni thinks here

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: > But update should be fine if it could have dict, or instance of the class as an argument?

Yeah I think it should be able to accept both if we go in this direction

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: > Ok that's good to know. Is that the case with Qt?

As I remember Qt native widgets des not emit a signal if you set the current value.

Comment (Czaki) [reply napari/utils/events/dataclass.py]: > We'd definitely have to fix these, but that might improve the overall codebase!

First, you need to dramatically increase the test codebase to add many integration tests between components. Otherwise, there will be many regressions that are hard to debug. 

And we do not start talking about user custom widgets. 

I would rather create a custom `property` like decorator which will do this check.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: My bad. I missed the connection to the emitter group.

Comment (Czaki) [reply napari/utils/events/dataclass.py]: removed

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: done

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: updated

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: So this is not a part of this PR and should be done in another PR?

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: I much prefer this. Do we need to add checks to see if the class has `events`? I think its still possible to create a non-evented dataclass, so we might want to add checks in

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py:567]: Very nice!

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: I'm fine with this for now, but we might want to revisit if there are performance issues

**Review by sofroniewn — COMMENTED** [review]
Ok this, is looking really good. I'd still prefer to remove the `deprecated_events` before merge and put that in a separate PR, as I'm not sure this will be how we want to handle deprecation.

We can leave the compare stuff in here for now as that doesn't change the API. We can change later if we want.

I like `asdict` and `update`, but want to hear from @jni too

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: done

What is the profit of use napari dataclass instead of one from the standard library when not using events?

**Comment by Czaki** [comment]
What did you think about add context manager to EmitterGroup and EventEmitter for blocking the signal? 

This allows writing short code which guarantees to raise the block even when an exception happens.

**Comment by sofroniewn** [comment]
> What did you think about add context manager to EmitterGroup and EventEmitter for blocking the signal?
> 
> This allows writing short code which guarantees to raise the block even when an exception happens.

Yes, we should do this. I think our event system already provides this as `with dataclass.events.blocker():` 
https://github.com/napari/napari/blob/v0.4.2/napari/utils/events/event.py#L590-L600

**Comment by Czaki** [comment]
> > What did you think about add context manager to EmitterGroup and EventEmitter for blocking the signal?
> > This allows writing short code which guarantees to raise the block even when an exception happens.
> 
> Yes, we should do this. I think our event system already provides this as `with dataclass.events.blocker():`
> [https://github.com/napari/napari/blob/v0.4.2/napari/utils/events/event.py#L590-L600](https://github.com/napari/napari/blob/v0.4.2/napari/utils/events/event.py?rgh-link-date=2020-12-01T21%3A31%3A00Z#L590-L600)

I need to learn more about napari events.

**Comment by sofroniewn** [comment]
> I need to learn more about napari events.

We vendored from vispy, so you can read more here http://vispy.org/util.html#module-vispy.util.event

**Comment by Czaki** [comment]
Any idea what could happen with macOS 3.8 test?

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: > What is the profit of use napari dataclass instead of one from the standard library when not using events?

One could use the `Properties` without the `Events`, but not clear how common that will be. We might want to change/ curious what @tlambert03 thinks

**Comment by sofroniewn** [comment]
> Any idea what could happen with macOS 3.8 test?

No, issue made in #1974 to discuss

File: napari/utils/events/dataclass.py
Comment (tlambert03) [inline napari/utils/events/dataclass.py]: this could also be if `isinstance(self, napari.utils.events.SupportsEvents)` ... which is [being used elsewhere](https://github.com/napari/napari/blob/79ff7b84d3b8dcf5940d3fca50db495d68c281d7/napari/utils/events/containers/_nested_list.py#L239) at lets us have one universal definition of what an "evented object" is.

**Comment by Czaki** [comment]
> No, issue made in #1974 to discuss

This is different case. here is segfault.

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [inline napari/utils/events/dataclass.py]: I'm a little nervous about this `self._hidden_event`. I assume it was necessary for some reason, but I don't immediately see why

Comment (Czaki) [reply napari/utils/events/dataclass.py]: Try to debugging macOS build fail.

**Comment by Czaki** [comment]
I do not understand why it fails previously. Maybe some fix on the runner allows passing

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: changed

Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: Thanks!!

**Comment by sofroniewn** [comment]
Thanks for these updates. The only remaining question as far as I am concerned is whether to include the special handling of `deprecated_events`. We can wait for @jni to weigh in before deciding, but I'd prefer that addition to come in it's own PR right before the first time we need it rather than right now as we might realize we need additional machinery to do a deprecation cleanly. I think if that was removed we could merge this right now.

**Comment by Czaki** [comment]
@sofroniewn I removed deprecated from this PR. Did you prefer the next PR or Issue?

File: napari/utils/events/dataclass.py
Comment (tlambert03) [reply napari/utils/events/dataclass.py]: Assuming this comparison functionality stays in this PR, can we rename this function to something that indicates its return value a little more?   something like "`is_equal`" instead of `compare` ... 
also, add a `Returns` section to the numpydocs and move the description of what is returned (and add a type) there:
```
Returns
-------
result : bool
    Returns ``True`` if ``v1`` and ``v2`` are equivalent.
    Dask arrays, and any two objects of different ``type`` will always return ``False``.
    If an exception is raised during comparison, returns ``False``.

Warns
-----
UserWarning
    If an exception is raised during comparison.
```

**Comment by sofroniewn** [comment]
> @sofroniewn I removed deprecated from this PR. Did you prefer the next PR or Issue?

Thanks, either is fine. You can start a new PR and we can continue discussion there

File: napari/utils/events/dataclass.py
Comment (sofroniewn) [reply napari/utils/events/dataclass.py]: Agreed, this would be nicer. With this change I'm then happy to leave in this PR, as it doesn't change current behavior and will fix all our warnings. It can be improved/ changed later.

**Review by sofroniewn — APPROVED** [review]
Ok great work @Czaki and thanks for bearing with us through all the comments. I'm approving now, but it would be nice to rename `compare` as @tlambert03 notes here https://github.com/napari/napari/pull/1966/files#r534417658

I will wait for that fix and in case @tlambert03 or @jni have any additional comments before merge, but think that we're basically there now!

**Comment by tlambert03** [comment]
Yeah thanks @Czaki 

I'm not opposed to merging this more or less as is, but I think it's very possible that, while this PR is definitely is going to help us avoid some uncaught exceptions,  it might also introduce a sneaky performance reduction.  So, if we go ahead and include this `compare` approach, i do think we should immediately open (and address quickly) an issue that actually benchmarks some things.  For instance, it _might_ be better to just try to compare things with `==` before doing all of this type checking.  There's tons of different scenarios, but _most_ of the properties in our models are rather simple types, and this will no doubt add overhead.

**Comment by Czaki** [comment]
In the previous version, there is a `if` for numpy arrays. 

And what You think about creating a few compare function and select (during class construction) which use base on a declared type. With left this one as most general?

**Comment by tlambert03** [comment]
> In the previous version, there is a if for numpy arrays.

yes, but that too was only recently added in #1910 and was not part of the original design ... And while some of these are going to be necessary, my point is just that we're dealing with a class here that will eventually affect almost every class in napari, so anytime we modify this class, we need to be extra careful about performance.

> And what You think about creating a few compare function and select (during class construction) which use base on a declared type. With left this one as most general?

I like this idea!  We could look for a class attribute like
```python

__equality_checks__ : Dict[str, Callable[[Any, Any], bool]]
```

... where the keys are attribute names and the `callable` is the equality check... and the default equality check is just `operator.eq`

**Comment by Czaki** [comment]
@tlambert03 What did you think about the current version? 

I decide to left simplified `is_equal`  to keep fallback ad prevent exception which could break the code. this is the only proposition to make code safer. If You think that `operator.eq` is better I will change it. 

I also decide to store only non-default options in the dictionary.

File: napari/utils/events/dataclass.py
Comment (jni) [inline napari/utils/events/dataclass.py:225]: Exceptions are relatively expensive in Python. We should try to avoid them. Here are some quick tests. arr1 is a (512, 512) dask array, arr2 and arr3 are numpy arrays of size (512,):

```python
In [10]: def check(a, b):
    ...:     try:
    ...:         return bool(a == b)
    ...:     except:
    ...:         return False
    ...:

In [13]: %timeit check(arr1, arr2)
7.73 ms ± 1.28 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [16]: %timeit check(arr2, arr3)
10 µs ± 328 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [18]: def check2(a, b):
    ...:     if type(a) != type(b): return False
    ...:     if hasattr(a, 'ndim') or hasattr(b, 'ndim'): return False
    ...:     try:
    ...:         return bool(a == b)
    ...:     except:
    ...:         return False
    ...:

In [19]: %timeit check2(arr1, arr2)
1.66 µs ± 75.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [20]: %timeit check2(arr2, arr3)
1.87 µs ± 139 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

File: napari/utils/events/dataclass.py
Comment (jni) [inline napari/utils/events/dataclass.py]: Could you add a comment above these lines about what compare_dict is going to be used for? These lines in isolation seem very arcane. Something like, "when checking for equality between two elements of a dataclass, rather than relying on built-in Python comparison, we will use the types then values in this compare_dict"

**Review by jni — APPROVED** [review]
Thanks @Czaki! This is very nice! I've made one additional documentation suggestion, as well as an implementation suggestion based on some quick benchmarks. Otherwise I think this is ready! 🎉

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py:225]: Something like this was implemented in original PR but I was requested to change its:
https://github.com/napari/napari/pull/1966#issuecomment-737449663

Currently, when types are properly declared NumPy arrays are compared using `np.array_equal` and ask arrays using `operator.is_` 

The pointed part of the code should be used only for types that could be easily compared but have a fallback mode to prevent program crashes on comparison failiture.

File: napari/utils/events/dataclass.py
Comment (Czaki) [reply napari/utils/events/dataclass.py]: Is this comment ok now?

File: napari/utils/events/dataclass.py
Comment (jni) [reply napari/utils/events/dataclass.py:225]: > Currently, when types are properly declared NumPy arrays are compared using np.array_equal and ask arrays using operator.is_

Ah, sorry, I missed this fact, and thought we were relying on the exception catching to compare NumPy arrays. Sorry about that!

**Comment by jni** [comment]
Thank you @Czaki!

**Comment by tlambert03** [comment]
thanks @Czaki!

**Comment by sofroniewn** [comment]
Just noticing that `asdict` and `update` are the same syntax on the `zarr.attrs` which is very nice! Will make for some nicer interoperability with the napari `evented_dataclass` https://zarr.readthedocs.io/en/stable/api/attrs.html

**Comment by sofroniewn** [comment]
They have a `put(d)` which actually might be more like our `update` .... we might want to move this adjust our syntax to match theirs? Thoughts @tlambert03 @Czaki, I'm open?

**Comment by tlambert03** [comment]
I don't know... both their `put(arg: dict)` and `update(*args, **kwargs)` are supported by the python `dict.update` function... so I'd stick with update (and make it more like the python mutable mapping than zarr):
```
In [1]: a = {}

In [2]: a.update(a = 4)

In [3]: a
Out[3]: {'a': 4}

In [4]: a.update({'b': 5})

In [5]: a
Out[5]: {'a': 4, 'b': 5}
```

**Comment by sofroniewn** [comment]
> I don't know... both their put(arg: dict) and update(*args, **kwargs) are supported by the python dict.update function... so I'd stick with update (and make it more like the python mutable mapping than zarr):

Ok that's fine, thanks for considering. I think as long as we're mindful of how other nearby packages are using similar names and we think that we're being overall consistent then that's good by me! We can always consider this again after we've gotten some more usage experience

**Pull request merged**
